qiskit/algorithms/time_evolution/variational/error_calculators/gradient_errors/error_calculator.py

```diff
@@ -0,0 +1,139 @@
+        self._backend = backend
+
+    def _bind_or_sample_operator(
+        self,
+        operator: OperatorBase,
+        operator_circuit_sampler: CircuitSampler,
+        param_dict: Dict[Parameter, Union[float, complex]],
+    ) -> OperatorBase:
+        """
+        Args:
+            operator: Operator to be bound with parameter values.
+            operator_circuit_sampler: CircuitSampler for the operator.
+            param_dict: Dictionary which relates parameter values to the parameters in the operator.
+        Returns:
+            Operator with all parameters bound.
+        """
+        # ⟨ψ(ω)|H^2|ψ(ω)〉
+        if operator_circuit_sampler:
+            operator = operator_circuit_sampler.convert(operator, params=param_dict)
+        else:
+            operator = operator.assign_parameters(param_dict)
+        operator = np.real(operator.eval())
+        return operator
+
+    @abstractmethod
+    def _calc_single_step_error(
+        self,
+        ng_res: Union[List, np.ndarray],
+        grad_res: Union[List, np.ndarray],
+        metric: Union[List, np.ndarray],
+        param_dict: Dict[Parameter, Union[float, complex]],
+    ) -> Tuple[int, Union[np.ndarray, complex, float], Union[Union[complex, float], Any],]:
+
+        """
+        Evaluate the l2 norm of the error for a single time step of VarQITE.
+        Args:
+            ng_res: dω/dt.
+            grad_res: 2Re⟨dψ(ω)/dω|H|ψ(ω)〉.
+            metric: Fubini-Study Metric.
+        Returns:
+            Square root of the l2 norm of the error.
+        """
+        pass
+
+    @abstractmethod
+    def _calc_single_step_error_gradient(
+        self,
+        ng_res: Union[List, np.ndarray],
+        grad_res: Union[List, np.ndarray],
+        metric: Union[List, np.ndarray],
+    ) -> float:
+        """
+        Evaluate the gradient of the l2 norm for a single time step of VarQRTE.
+        Args:
+            ng_res: dω/dt.
+            grad_res: -2Im⟨dψ(ω)/dω|H|ψ(ω)〉.
+            metric: Fubini-Study Metric.
+        Returns:
+            Square root of the l2 norm of the error.
+        """
+        pass
+
+    def _inner_prod(self, x: Iterable, y: Iterable) -> Union[np.ndarray, np.complex, np.float]:
+        """
+        Compute the inner product of two vectors.
+        Args:
+            x: vector.
+            y: vector.
+        Returns:
+            Inner product of x and y.
+        """
+        return np.matmul(np.conj(np.transpose(x)), y)
+
+    def _validate_epsilon_squared(self, eps_squared: float) -> float:
+        """
+        Sets an epsilon provided to 0 if it is small enough and negative (smaller than 1e-3 in
+        absolute value).
+        Args:
+            eps_squared: Value to be validated.
+        Returns:
+            Modified (or not) value provided.
+        Raises:
+            Warning if the value provided is too large.
+        """
+        if eps_squared < 0:
+            if np.abs(eps_squared) < 1e-3:
```

Comment:
This seems a bit sketchy to do, especially without a warning or log. Why is a value of `1e-3` the right cutoff? And why is a negative epsilon invalid (maybe we should raise a clearer error in this case)?

Reply:
Yeah, I agree with everything you pointed out. I cannot comment on the choice of `1e-3` because it comes from the prototype code. I think I discussed changing a warning to an error with Christa at some point. @Zoufalc any motivation for having this code from your side?

Reply:
Thanks for pointing this out.
It can easily happen that we get very small negative values that simply arise from a numerical instability. `1e-3` actually seems too large for me. This should be somewhere around `1e-6` or `1e-8`.
@dlasecki As mentioned before, I am perfectly fine if you want to change the warning into an error.